### PR TITLE
Move to Pillow 10.x

### PR DESCRIPTION
**Pull request by hostcc**
# What does this implement/fix?
Moving to recent versions of Pillow Python package has an immediate benefit of closing [security
issue](https://security.snyk.io/vuln/SNYK-PYTHON-PILLOW-5918878) has been reported recently.

**NOTE** The change has only been tested for compilation, but not on a real hardware since I don't have a display at the moment.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** N/A

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** https://github.com/esphome/esphome-docs/pull/3243

## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
# Example config.yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: requirements_optional.txt [hunk 1]
```diff
@@ -1,3 +1,3 @@
-pillow>4.0.0,<10.0.0
+pillow>=10.0.0
```
Comment (frenck) [inline requirements_optional.txt]: Pillow 10.0.0 contains a high-severity vulnerability.

```suggestion
pillow>=10.0.1
```

Comment (hostcc) [reply requirements_optional.txt]: Sharp eye, sir - was targeting to address that exact case, thanks!

File: esphome/components/font/__init__.py [hunk 2]
```diff
@@ -67,18 +67,13 @@ def validate_pillow_installed(value):
     except ImportError as err:
         raise cv.Invalid(
             "Please install the pillow python package to use this feature. "
-            '(pip install pillow">4.0.0,<10.0.0")'
+            '(pip install pillow">=10.0.0")'
```
Comment (nielsnl68) [inline esphome/components/font/__init__.py]:
``` suggestion
            '(pip install pillow">10.0.0")'
```
when  10.0.0 has that vulnerability as @frenck said, should we not disallow that version to be installed as well?

File: esphome/components/font/__init__.py [hunk 3]
```diff
@@ -67,18 +67,13 @@ def validate_pillow_installed(value):
...
         raise cv.Invalid(
-            "Please update your pillow installation to at least 4.0.x. "
-            '(pip install pillow">4.0.0,<10.0.0")'
-        )
-    if version.parse(PIL.__version__) >= version.parse("10.0.0"):
-        raise cv.Invalid(
-            "Please downgrade your pillow installation to below 10.0.0. "
-            '(pip install pillow">4.0.0,<10.0.0")'
+            "Please update your pillow installation to at least 10.0.x. "
+            '(pip install pillow">=10.0.0")'
```
Comment (nielsnl68) [inline esphome/components/font/__init__.py]: same as above.

File: esphome/components/font/__init__.py [hunk 4]
```diff
@@ -67,18 +67,13 @@ def validate_pillow_installed(value):
     except ImportError as err:
         raise cv.Invalid(
             "Please install the pillow python package to use this feature. "
-            '(pip install pillow">4.0.0,<10.0.0")'
+            '(pip install pillow">=10.0.0")'
         ) from err
 
-    if version.parse(PIL.__version__) < version.parse("4.0.0"):
+    if version.parse(PIL.__version__) < version.parse("10.0.0"):
```
Comment (nielsnl68) [inline esphome/components/font/__init__.py]: same as above

**Review by nielsnl68 — CHANGES_REQUESTED** [review]
Thanks for updating this code. Based on @frenck 's suggestion about the vulnerability of 10.0.0, i have some suggested changes as well.

File: esphome/components/font/__init__.py (hunk 2 above)
Comment (hostcc) [reply esphome/components/font/__init__.py]: Makes a perfect sense - thanks! addressed with eb35958d

**Review by nielsnl68 — APPROVED** [review]

File: requirements_optional.txt [hunk 5]
```diff
@@ -1,3 +1,3 @@
-pillow>4.0.0,<10.0.0
+pillow>=10.0.1
```
Comment (jesserockz) [inline requirements_optional.txt]: I actually want to move these to be exact versions, so may as well make that here
```suggestion
pillow==10.0.1
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: requirements_optional.txt (hunk 5 above)
Comment (hostcc) [reply requirements_optional.txt]: @jesserockz , I agree pinning dependencies to exact versions is great, I initially followed same semantics as others under `requirements_optional.txt. Will update the PR shortly.

**Comment by jesserockz** [comment]
CI errors unrelated, fixed with #5501

File: esphome/components/font/__init__.py [hunk 6]
```diff
@@ -67,18 +67,13 @@ def validate_pillow_installed(value):
     except ImportError as err:
         raise cv.Invalid(
             "Please install the pillow python package to use this feature. "
-            '(pip install pillow">4.0.0,<10.0.0")'
+            '(pip install pillow">=10.0.1")'
```
Comment (jesserockz) [inline esphome/components/font/__init__.py]:
```suggestion
            '(pip install "pillow==10.0.1")'
```

File: esphome/components/font/__init__.py [hunk 7]
```diff
@@ -67,18 +67,13 @@ def validate_pillow_installed(value):
...
         ) from err
 
-    if version.parse(PIL.__version__) < version.parse("4.0.0"):
+    if version.parse(PIL.__version__) < version.parse("10.0.1"):
```
Comment (jesserockz) [inline esphome/components/font/__init__.py]:
```suggestion
    if version.parse(PIL.__version__) != version.parse("10.0.1"):
```

File: esphome/components/font/__init__.py [hunk 8]
```diff
@@ -67,18 +67,13 @@ def validate_pillow_installed(value):
...
         raise cv.Invalid(
-            "Please update your pillow installation to at least 4.0.x. "
-            '(pip install pillow">4.0.0,<10.0.0")'
-        )
-    if version.parse(PIL.__version__) >= version.parse("10.0.0"):
-        raise cv.Invalid(
-            "Please downgrade your pillow installation to below 10.0.0. "
-            '(pip install pillow">4.0.0,<10.0.0")'
+            "Please update your pillow installation to at least 10.0.1. "
+            '(pip install pillow">=10.0.1")'
```
Comment (jesserockz) [inline esphome/components/font/__init__.py]:
```suggestion
            "Please update your pillow installation to 10.0.1. "
            '(pip install "pillow==10.0.1")'
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

**Review by jesserockz — APPROVED** [review]

**Pull request merged**